memcached-operator/pkg/controller/memcached/memcached_controller.go

```diff
@@ -187,17 +174,18 @@ func (r *ReconcileMemcached) Reconcile(request reconcile.Request) (reconcile.Res
 	return reconcile.Result{}, nil
 }
 
-// deploymentForMemcached returns a memcached Deployment object
-func (r *ReconcileMemcached) deploymentForMemcached(m *cachev1alpha1.Memcached) *appsv1.Deployment {
-	ls := labelsForMemcached(m.Name)
+// deploymentForMemcached mutates the passed in Deployment reference to the expected state
+func deploymentForMemcached(m *cachev1alpha1.Memcached, reqLogger logr.Logger, deployment *appsv1.Deployment) {
+	name := "memcached"
+	image := "memcached:1.4.36-alpine"
 	replicas := m.Spec.Size
+	ls := labelsForMemcached(m.Name)
+	command := []string{"memcached", "-m=64", "-o", "modern", "-v"}
+	ports := []corev1.ContainerPort{{ContainerPort: 11211, Name: "memcached", Protocol: "TCP"}}
 
-	dep := &appsv1.Deployment{
-		ObjectMeta: metav1.ObjectMeta{
-			Name:      m.Name,
-			Namespace: m.Namespace,
-		},
-		Spec: appsv1.DeploymentSpec{
+	if deployment.ObjectMeta.CreationTimestamp.IsZero() {
```

Comment:
Why add this if and the checks if these impl shows ensure for ALL specs that the original should not be changed for example? Could not so to be simplified return the object with the CR.spec values always?

Reply:
There are a lot of questions here, I'm not sure I understand all of them but I'll try to answer as much as I can:

- This check to see if the object is created is really for 2 main reasons:
  - create the initial structure of the object -- especially things like the Container element in the Containers list, so that future invocations of the function can just expect it to be there (although now that I say that out loud, that might be a bit naive :thinking:)
  - There are things like the `Selector` in the Spec here, which appear to be only modifiable on initial creation, rather than updatable every time. You'll notice that Selector is populated here, but not reset later.

- I'm not sure if the only goal is only to maintain values that may have been specified in the primary resource (Memcached CR here), is that what you're asking about? Even if something is not specified (or even exposed) through that, I think we should strive to keep it in the state that this controller decides, rather than letting users accidentally (or on purpose) mess things up.

I know there are other questions there too, but I don't understand them enough to attempt to answer. Could you give an example of what you mean? Also, I've made some changes just now, including adding some short explanatory comments -- maybe those questions are answered now?

Reply:
H @grdryn,

It is not mandatory and will not the best approach in all scenarios/for all business ensure that the Deployment and/or any resource is kept as its original state. It is really up the business rules. 

For example, we can CRUD new resources in the case of the replicas of a Deployment be increased. So, do changes that can lead to the users that it should or could NOT be done is something that concerns me.